// zkwasm-service-cli/dist/commands/queryTask.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.builder = exports.desc = exports.command = void 0;
const query_1 = require("../query");
exports.command = "querytask";
exports.desc = "Query task";
const builder = (yargs) => {
    return yargs.option("t", {
        alias: "tx",
        describe: "transaction hash",
        type: "string",
        demandOption: "The transaction hash is required",
    });
};
exports.builder = builder;
const handler = (argv) => __awaiter(void 0, void 0, void 0, function* () {
    console.log("Querying task...");
    yield (0, query_1.queryTask)(argv.t, argv.r);
});
exports.handler = handler;
